backupccl: target offline importing tables if flag is set

msbutler opened the pull request:
Historically, offline tables were excluded from backups (e.g. restoring or
importing tables) because incremental backups could not reason correctly about
their MVCC history. Now that all within tenant operations are now MVCC, BACKUP
can begin incrementally backing up offline tables.

Independent of the MVCC history thread, https://github.com/cockroachdb/cockroach/pull/83915 allowed backup to implicitly
target _all_ offline descriptors (e.g. database, schema, table, type) to
account for a class of compound online schema changes. This is a regression
from the expected backup behavior and will get reverted in future work.

This patch ensures that offline tables with importing data are not targetted
iff an external flag is set.

This patch merely refactors the internal targeting logic in
backupresolver.ResolveTargetsToDescriptors, and does not change how backups
behave. In other words, after this patch, no offline descriptors will get
targeted, as all callers set the function's backupImports flag to false. A
future PR will change these callsites.

Informs https://github.com/cockroachdb/cockroach/issues/76722

Release note: none

Release justification: low risk change to a codepath that does not currently get called in production

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85979)
<!-- Reviewable:end -->

msbutler commented:
RFAL! 

only the latest commit is relevant, the current test failures are unrelated (https://github.com/cockroachdb/cockroach/pull/85350#issuecomment-1212459421)

msbutler commented on pkg/ccl/backupccl/targets.go:
```diff
@@ -257,6 +257,11 @@ func fullClusterTargets(
 				fullClusterDBs = append(fullClusterDBs, dbDesc)
 			}
 		case catalog.TableDescriptor:
```
before this diff and Marius's, we were including offline tables in cluster backups. Was that intended? @dt

dt replied:
I don't think so (cc @adityamaru in case I'm forgetting something).

adityamaru replied:
Not that I can recall, unless the presence of the descriptors was something that affected our ability to roll back the import on restore?

dt replied:
Ah, yeah, that was probably it -- we probably included the offline tables because restored jobs referred to them.

msbutler replied:
So, since we're no longer backing up jobs in cluster restores, are we cool with only backing up offline tables, if they're apart of an import rollback? But maybe that will screw with other offline descriptors that have in-progress schema change jobs?

Maybe I'll keep offline tables in the cluster for now, and audit this logic during stability.

msbutler replied:
will remove offline tables after andrew removes offline state from schema change job

msbutler commented on pkg/ccl/backupccl/testdata/backup-restore/backup-dropped-descriptors-declarative:
```diff
@@ -110,31 +111,35 @@ SELECT orig->'type'->'name', orig->'type'->'state' FROM tbls WHERE id = 110 OR i
+typ
+_typ
+s
 
```
Perhaps this PR doesn't go far enough: as you see above,`typ` and `_typ` are offline type descriptors, and `s` is an offline schema and both are included in the (cluster) backup, because as of https://github.com/cockroachdb/cockroach/pull/83915, the drop job moves these guys briefly into an offline state. 

That _seems_ fine, as long as the online schema changer follows through and drops these descriptors in RESTORE. In other words, I _think_ the restoring cluster should not see schema s. But below, the test _expects_ to see offline schema s in the restored database. That seems wrong? Or maybe the backed up offline schema is live in the restored cluster because the restore job rolls back the drop? But is that also incorrect???

postamar replied:
Types and schemas can't reference tables so it should be OK, however, tables can. I suggest you add a test case (if there isn't one already) with views which reference a type and a table that are also being dropped, and see whether the schema change chugs along post-restore.

msbutler replied:
Why do you think this is OK?

`typ` (type desc)  and `s` (schema desc) were in the middle of getting dropped when the backup occurred. This test suggests that these descriptors are now _online_ in the restored cluster.  I think these descriptors should not exist in the restored cluster, as they were getting dropped in the when the backup was taken. But maybe i'm misunderstanding something?

dt replied:
I thought you can't drop a table if a view depends on it?

msbutler replied:
resolved offline: andrew will refactor the schema changer to not use the offline state. Down the road, restore/import will create a new descriptor state for IMPORTING and RESTORING.

msbutler committed:
backupccl: target offline importing tables if flag is set

Historically, offline tables were excluded from backups (e.g. restoring or
importing tables) because incremental backups could not reason correctly about
their MVCC history. Now that all within tenant operations are now MVCC, BACKUP
can begin incrementally backing up offline tables.

Independent of the MVCC history thread, #83915 allowed backup to implicitly
target _all_ offline descriptors (e.g. database, schema, table, type) to
account for a class of compound online schema changes. This is a regression
from the expected backup behavior and will get reverted in future work.

This patch ensures that offline tables with importing data are not targetted
iff an external flag is set.

This patch merely refactors the internal targeting logic in
backupresolver.ResolveTargetsToDescriptors, and does not change how backups
behave. In other words, after this patch, no offline descriptors will get
targeted, as all callers set the function's backupImports flag to false. A
future PR will change these callsites.

Informs #76722

Release note: none

Release note (<category, see below>): <what> <show> <why>

msbutler commented:
This new diff will only target in-progress imports iff an external flag is set. It doesn't try to deal with any other kinds of offline descriptors.

Realized that I needed to modify targets.go in order to RESTORE TABLE with an in-flight import correctly.

@ajwerner We can delay this PR until after you refactor the schema changer to not depend on the offline state. Ideally, I can slip this in, so I can avoid doing duplicate implementation work.

msbutler commented:
Closing for now. We can figure out how to filter offline descriptors in restore properly without modify DescMatchingTargets.
